### PR TITLE
Added cluster setup using go processes

### PR DESCRIPTION
**Pull request by arindamnayak**
Added cluster setup using go processes (vttablet, vtctld, vtgate etc.. processes). 
Added clustertest to verify if cluster setup is working.

### CODE REVIEW

File: dev.env [hunk 1]
```diff
@@ -73,6 +73,17 @@ PATH=$(prepend_path "$PATH" "$VTROOT/dist/chromedriver")
 PATH=$(prepend_path "$PATH" "$VTROOT/dist/node/bin")
 export PATH
 
+# Etcd path.
```
Comment (deepthi) [inline dev.env:76]: Why is this needed?

File: go/test/endtoend/cluster/cluster_process.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+	// background executable processes
+	topoProcess   EtcdProcess
+	vtctldProcess VtctldProcess
+	vtgateProcess VtgateProcess
+
+	nextPortForProcess int
+}
+
+// Keyspace Cluster accepts keyspace to launch it
```
Comment (deepthi) [inline go/test/endtoend/cluster/cluster_process.go]: add a `:` after `Keyspace`

File: go/test/endtoend/cluster/cluster_process.go [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+type Keyspace struct {
+	Name      string
+	SQLSchema string
+	VSchema   string
+	Shards    []Shard
+}
+
+// Shard Keyspace need Shard to manage its lifecycle
```
Comment (deepthi) [inline go/test/endtoend/cluster/cluster_process.go]: `Shard with associated vttablets`

File: go/test/endtoend/cluster/cluster_process.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+type Shard struct {
+	Name      string
+	Vttablets []Vttablet
+}
+
+// Vttablet Shard need vttablet to manage the setup and teardown
```
Comment (deepthi) [inline go/test/endtoend/cluster/cluster_process.go]: `Vttablet stores the properties needed to start a vttablet process`

File: go/test/endtoend/cluster/cluster_process.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+	}
+	cluster.TopoPort = cluster.GetAndReservePort()
+	cluster.topoProcess = *EtcdProcessInstance(cluster.TopoPort, cluster.Hostname)
+	println(fmt.Sprintf("Starting etcd server on port : %d", cluster.TopoPort))
+	if err = cluster.topoProcess.Setup(); err != nil {
+		println(err.Error())
+		return
+	}
+
+	println("Creating topop dirs")
```
Comment (deepthi) [inline go/test/endtoend/cluster/cluster_process.go]: typo: topop -> topo

File: go/test/endtoend/cluster/cluster_process.go [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+	vtctldProcess VtctldProcess
+	vtgateProcess VtgateProcess
+
+	nextPortForProcess int
+}
+
+// Keyspace Cluster accepts keyspace to launch it
+type Keyspace struct {
+	Name      string
+	SQLSchema string
```
Comment (deepthi) [inline go/test/endtoend/cluster/cluster_process.go]: This field should be called `SchemaSQL`  instead.

File: go/test/endtoend/cluster/cluster_process.go [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+			if err = tablet.vttabletProcess.Setup(); err != nil {
+				println(err.Error())
+				return
+			}
+
+			shard.Vttablets = append(shard.Vttablets, *tablet)
+		}
+
+		// Make 1st tablet as master shard
```
Comment (deepthi) [inline go/test/endtoend/cluster/cluster_process.go]: `Make first tablet as master`

File: go/test/endtoend/cluster/cluster_process.go [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+		if err = cluster.VtctlclientProcess.InitShardMaster(keyspace.Name, shardName, cluster.Cell, shard.Vttablets[0].TabletUID); err != nil {
+			println(err.Error())
+			return
+		}
+
+		// Apply SQLSchema
```
Comment (deepthi) [inline go/test/endtoend/cluster/cluster_process.go]: `Apply schema sql`

File: go/test/endtoend/cluster/etcd_process.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+	return false
+}
+
+func (etcd *EtcdProcess) removeTopoDirectories(Cell string) {
+	etcd.ManageTopoDir("rmdir", "/vitess/global")
+	etcd.ManageTopoDir("rmdir", "/vitess/"+Cell)
+}
+
+// ManageTopoDir creates global and zone in etcd2
+func (etcd *EtcdProcess) ManageTopoDir(command string, directory string) error {
```
Comment (deepthi) [inline go/test/endtoend/cluster/etcd_process.go:135]: `command` is not being used here, instead it is always `mkdir`. If the only use of this function is to create directories, then it should be renamed to `CreateTopoDir`, and the fist argument (`command`) can be dropped.

File: go/test/endtoend/cluster/etcd_process.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	tmpProcess := exec.Command(
```
Comment (deepthi) [inline go/test/endtoend/cluster/etcd_process.go:136]: you can save time by naming such variables with single letters - like `p` for process. That is acceptable.

File: go/test/endtoend/cluster/vtctlclient_process.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	defer os.Remove(file.Name())
+
+	_, _ = file.WriteString(SQL)
+	_ = file.Close()
+	tmpProcess := exec.Command(
+		vtctlclient.Binary,
+		"-server", vtctlclient.Server,
+		"ApplySchema",
+		"-sql-file", file.Name(),
```
Comment (deepthi) [inline go/test/endtoend/cluster/vtctlclient_process.go]: I don't think you need to create a file and pass it. `ApplySchema` has an option `-sql` where you can directly provide the schema sql string.

File: go/test/endtoend/cluster/vtctlclient_process.go [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+		return
+	}
+	defer os.Remove(file.Name())
+
+	_, _ = file.WriteString(JSON)
+	_ = file.Close()
+	tmpProcess := exec.Command(
+		vtctlclient.Binary,
+		"-server", vtctlclient.Server,
+		"ApplyVSchema",
```
Comment (deepthi) [inline go/test/endtoend/cluster/vtctlclient_process.go]: similar option here: `-vschema`

File: go/test/endtoend/cluster/vttablet_process.go [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+	"os/exec"
+	"path"
+	"strings"
+	"syscall"
+	"time"
+
+	"vitess.io/vitess/go/vt/log"
+)
+
+// VttabletProcess is a generic handle for a running vtctld .
```
Comment (deepthi) [inline go/test/endtoend/cluster/vttablet_process.go]: `a running vttablet`

File: go/test/endtoend/cluster/vttablet_process.go [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+			return fmt.Errorf("process '%s' exited prematurely (err: %s)", vttablet.Name, err)
+		default:
+			time.Sleep(300 * time.Millisecond)
+		}
+	}
+
+	return fmt.Errorf("process '%s' timed out after 60s (err: %s)", vttablet.Name, <-vttablet.exit)
+}
+
+// WaitForStatus function checks if vtctld process is up and running
```
Comment (deepthi) [inline go/test/endtoend/cluster/vttablet_process.go]: vtctld -> vttablet

File: go/test/endtoend/cluster/vttablet_process.go [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+		err := json.Unmarshal(respByte, &resultMap)
+		if err != nil {
+			panic(err)
+		}
+		return resultMap["TabletStateName"] == status
+	}
+	return false
+}
+
+// TearDown shutdowns the running vttablet service
```
Comment (deepthi) [inline go/test/endtoend/cluster/vttablet_process.go]: shutdowns -> shuts down

File: go/test/endtoend/cluster/vttablet_process.go [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+		return err
+
+	case <-time.After(10 * time.Second):
+		vttablet.proc.Process.Kill()
+		vttablet.proc = nil
+		return <-vttablet.exit
+	}
+}
+
+// VttabletProcessInstance returns a VtctlProcess handle for vtctl process
```
Comment (deepthi) [inline go/test/endtoend/cluster/vttablet_process.go]: Fix this comment.

File: go/test/endtoend/cluster/vtgate_process.go [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+		//	println("VTGate API Response: Key = " + key + ", value = " + fmt.Sprintf("%v", value))
+		//}
+		//println(string(respByte))
+		//return resultMap["TabletStateName"] == "NOT_SERVING"
+		return true
+	}
+	return false
+}
+
+// TearDown shutdowns the running vttablet service
```
Comment (deepthi) [inline go/test/endtoend/cluster/vtgate_process.go]: shutdowns -> shuts down
vttablet -> vtgate

File: go/test/endtoend/clustertest/main_test.go [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+import (
+	"flag"
+	"net/http"
+	"os"
+	"testing"
+	"vitess.io/vitess/go/test/endtoend/cluster"
+)
+
+var (
+	ClusterInstance *cluster.LocalProcessCluster
```
Comment (deepthi) [inline go/test/endtoend/clustertest/main_test.go]: Are any of these variables used outside this package? Or expected to be used that way?
If not, they should start with lowercase.

File: go/test/endtoend/cluster/cluster_process.go [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+
+	// background executable processes
+	mysqlctlProcess MysqlctlProcess
+	vttabletProcess VttabletProcess
+}
+
+// StartTopo starts topology server
+func (cluster *LocalProcessCluster) StartTopo() (err error) {
+	if cluster.Cell == "" {
+		cluster.Cell = "zone1"
```
Comment (deepthi) [inline go/test/endtoend/cluster/cluster_process.go]: Use a constant instead of hardcoding.

**Review by deepthi — CHANGES_REQUESTED** [review]
Nice work! I have requested a number of small changes.
Two things to note:
* You will need to add the ability to execute commands using `vtctl` or `vtctlclient` before you can write the first _real_ test case.
* Right now all log goes to stdout / stderr. That's fine, but pretty soon we will want to send the log for each process to a different file.

File: go/test/endtoend/cluster/mysqlctl_process.go [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+	"path"
+)
+
+// MysqlctlProcess is a generic handle for a running mysqlctl command .
+// It can be spawned manually
+type MysqlctlProcess struct {
+	Name         string
+	Binary       string
+	LogDirectory string
+	TabletUID    int
```
Comment (deepthi) [inline go/test/endtoend/cluster/mysqlctl_process.go:33]: Where will the actual mysql data and log be stored?

File: dev.env (hunk 1 above)
Comment (arindamnayak) [reply dev.env:76]: Currently `dev.env` does not set path for binary directory of `etcd`. That is why I added here to set the `etcd` path, so that go process can directly spawn `etcd` process.

File: go/test/endtoend/cluster/mysqlctl_process.go (hunk 20 above)
Comment (arindamnayak) [reply go/test/endtoend/cluster/mysqlctl_process.go:33]: Mysql data is stored at : `$VTDATAROOT/vt_${tablet_uid}/data`. And mysql log is stored at `$VTDATAROOT/tmp`

File: go/test/endtoend/clustertest/main_test.go (hunk 18 above)
Comment (arindamnayak) [reply go/test/endtoend/clustertest/main_test.go]: Yes, there are being using in individual testcases e.g. `vtgate_test.go` located in the same package.

File: go/test/endtoend/cluster/etcd_process.go (hunk 9 above)
Comment (arindamnayak) [reply go/test/endtoend/cluster/etcd_process.go:135]: During `TearDown`, we need to clean up the TopoDir, there the command used as `rmdir` - https://github.com/planetscale/vitess/blob/tal_vtgate_cluster/go/test/endtoend/cluster/etcd_process.go#L130. That is why I need to pass the `command`.

File: go/test/endtoend/cluster/vtctlclient_process.go (hunk 11 above)
Comment (arindamnayak) [reply go/test/endtoend/cluster/vtctlclient_process.go]: Changed code to use `SQL` directly instead of file.

File: go/test/endtoend/cluster/vtctlclient_process.go (hunk 12 above)
Comment (arindamnayak) [reply go/test/endtoend/cluster/vtctlclient_process.go]: Changed code to use `JSON` directly instead of file.

**Comment by arindamnayak** [comment]
> Nice work! I have requested a number of small changes.
> Two things to note:
> 
> * You will need to add the ability to execute commands using `vtctl` or `vtctlclient` before you can write the first _real_ test case.
> * Right now all log goes to stdout / stderr. That's fine, but pretty soon we will want to send the log for each process to a different file.

Thanks @deepthi , I have implemented all changes you requested, also answered to the questions. I have added method to execute any custom command for `vtctlclient`. I have also added log statements, so that all standard log command line params will work e.g. `go test -log_dir=/opt/cluster_test  -stderrthreshold=INFO -alsologtostderr=true` .

**Comment by arindamnayak** [comment]
I have also ported existing vtgate test to use current cluster setup. But vstream_test.go is not working, as, I am not able to register grpc dialer. When it runs from vtgate package, that dialer was registered automatically.

File: go/test/endtoend/cluster/etcd_process.go (hunk 9 above)
Comment (deepthi) [reply go/test/endtoend/cluster/etcd_process.go:135]: Then there is a bug in ManageTopoDir because that is always calling etcdctl with mkdir.

File: go/test/endtoend/cluster/mysqlctl_process.go (hunk 20 above)
Comment (deepthi) [reply go/test/endtoend/cluster/mysqlctl_process.go:33]: What happens when tests are run in parallel? Is tablet_uid unique? otherwise, you will have multiple tests using the same data dir.

File: go/test/endtoend/clustertest/main_test.go (hunk 18 above)
Comment (deepthi) [reply go/test/endtoend/clustertest/main_test.go]: If it is all in the same package, please change them to start with lowercase, which means they can't be used outside the package.

File: go/test/endtoend/cluster/vtctlclient_process.go [hunk 21]
```diff
@@ -73,25 +64,27 @@ func (vtctlclient *VtctlClientProcess) ApplySchema(Keyspace string, SQL string)
...
 	return tmpProcess.Run()
 }
 
+// ExecuteCommand executes any vtctlclient command
+func (vtctlclient *VtctlClientProcess) ExecuteCommand(args ...string) (err error) {
+	tmpProcess := exec.Command(
+		vtctlclient.Binary,
+		args...,
+	)
+	print(fmt.Sprintf("ApplyVSchema with arguments %v", strings.Join(tmpProcess.Args, " ")))
```
Comment (deepthi) [inline go/test/endtoend/cluster/vtctlclient_process.go]: Change this log, it is not always ApplyVSchema.

**Comment by deepthi** [comment]
@arindamnayak I have a couple of responses / new comments.

File: go/test/endtoend/cluster/mysqlctl_process.go (hunk 20 above)
Comment (arindamnayak) [reply go/test/endtoend/cluster/mysqlctl_process.go:33]: Yes, tablet_uid is unique. The data directory is different for each of tablets.

File: go/test/endtoend/cluster/etcd_process.go (hunk 9 above)
Comment (arindamnayak) [reply go/test/endtoend/cluster/etcd_process.go:135]: This will be fixed in next commit.

File: go/test/endtoend/cluster/vtctlclient_process.go (hunk 21 above)
Comment (arindamnayak) [reply go/test/endtoend/cluster/vtctlclient_process.go]: I already changed this in some earlier commit.

**Comment by arindamnayak** [comment]
> @arindamnayak I have a couple of responses / new comments.

@deepthi , I have updated code, please have a look.

**Review by deepthi — APPROVED** [review]

**Comment by deepthi** [comment]
Do not merge into planetscale/vitess:master

**Comment by ajeetj** [comment]
Closing as we have created PR against vitess.io
